src/python/WMCore/MicroService/MSPileup/MSPileupError.py

```diff
@@ -0,0 +1,110 @@
+"""
+File       : MSPileupError.py
+Author     : Valentin Kuznetsov <vkuznet AT gmail dot com>
+Description: MSPileupError represents MSPileupError module
+"""
+
+# system modules
+import json
+
+
+# MSPileup error codes
+MSPILEUP_GENERIC_ERROR = 1
+MSPILEUP_INVALID_ERROR = 2
+MSPILEUP_DUPLICATE_ERROR = 3
+MSPILEUP_NOTFOUND_ERROR = 4
+MSPILEUP_DATABASE_ERROR = 5
+MSPILEUP_UNIQUE_ERROR = 6
+
+
+class MSPileupError(object):
+    """
+    MSPileupError class represents generic MSPileup error
+    """
+    def __init__(self, data, msg=""):
+        """
+        Constructor of MSPileupError class
+        """
+        self.data = data
+        self.assign(msg)
+
+    def error(self):
+        """
+        Representation of MSPileupError
+
+        :return: MSPileupError representation
+        """
+        edict = {'data': self.data, 'message': self.msg, 'code': self.code, 'error': 'MSPileupError'}
+        return edict
+
+    def __str__(self):
+        """
+        String representation of MSPileupError
+
+        :return: human readable MSPileupError representation
+        """
+        return json.dumps(self.error(), indent=4)
+
+    def assign(self, msg="", code=0):
+        """
+        Assign proper message and error codes
+        :param msg: string
+        :param code: int
+        :return: None
+        """
+        if msg:
+            self.msg = msg
+        else:
+            self.msg = 'Generic MSPileup error'
+        if code > 0:
+            self.code = code
+        else:
+            self.code = MSPILEUP_GENERIC_ERROR
+
+    def GenericError(self, msg=""):
+        """
+        creates and MSPileup exception
+        """
+        self.assign(msg)
+        return self.error()
+
+    def InvalidData(self):
+        """
+        creates and MSPileup exception
+        """
+        self.assign(msg="invalid data", code=MSPILEUP_INVALID_ERROR)
+        return self.error()
+
+    def DuplicateDocument(self):
+        """
+        DUplicatte document MSPileup exception
+        """
+        msg = self.msg if self.msg else "duplicate MSPileup document found"
+        self.assign(msg=msg, code=MSPILEUP_DUPLICATE_ERROR)
+        return self.error()
+
+    def NoKeyFound(self):
+        """
+        No document found MSPileup exception
+        """
+        msg = self.msg if self.msg else "no document found"
+        self.assign(msg=msg, code=MSPILEUP_NOTFOUND_ERROR)
+        return self.error()
+
+    def DatabaseError(self, err=None):
+        """
+        Database error MSPileup exception
+        """
+        msg = "Dataset error"
+        if err:
+            msg += f" err={err}"
+        self.assign(msg=msg, code=MSPILEUP_DATABASE_ERROR)
+        return self.error()
+
+    def UniqueConstrainError(self):
```

Comment:
Even though these methods are supposed to be an specific exception, I wonder whether we should stick to the .pylintrc convention and start their names with a lower case char. What do you think?

Reply:
Alan, here and in general I follow python style convention. Said that, in python all exceptions are capital case, in other 3rd party modules, e.g. pymongo, they are in capital case, etc. Therefore, I think we should be consistent with python style. If you still insist on making them lower case please say so and I'll change that, I just wanted to reply to your comment with reason I had in mind and let you think about it one more time.

Reply:
done, new code contains child exception classes